crates/sargon/src/system/sargon_os/pre_authorization/sargon_os_create_subintent.rs

```diff
@@ -0,0 +1,250 @@
+        subintent_manifest: SubintentManifest,
+        expiration: DappToWalletInteractionSubintentExpiration,
+    ) -> Result<Subintent> {
+        let network_id = self.current_network_id()?;
+        let gateway_client = GatewayClient::new(
+            self.clients.http_client.driver.clone(),
+            network_id,
+        );
+        let current_epoch = gateway_client.current_epoch().await?;
+
+        // Calculate the seconds until the expiration of the subintent.
+        let expiry_time_from_now_in_seconds =
+            self.expiry_time_from_now_in_seconds(expiration);
+        if expiry_time_from_now_in_seconds == 0 {
+            return Err(CommonError::SubintentExpired);
+        }
+
+        // Calculate header ranges
+        let end_ranges = self.calculate_end_ranges(
+            current_epoch,
+            expiry_time_from_now_in_seconds,
+        );
+
+        // Build header
+        let header = IntentHeaderV2 {
+            network_id,
+            start_epoch_inclusive: current_epoch,
+            end_epoch_exclusive: end_ranges.0,
+            min_proposer_timestamp_inclusive: None,
+            max_proposer_timestamp_exclusive: Some(end_ranges.1),
+            intent_discriminator,
+        };
+
+        // Build subintent
+        Subintent::new(header, subintent_manifest, MessageV2::None)
+    }
+
+    /// Returns the seconds until the expiration of the subintent.
+    fn expiry_time_from_now_in_seconds(
+        &self,
+        expiration: DappToWalletInteractionSubintentExpiration,
+    ) -> u64 {
+        match expiration {
+            DappToWalletInteractionSubintentExpiration::AtTime(at_time) => {
+                at_time
+                    .unix_timestamp_seconds
+                    .duration_since(Timestamp::now_utc())
+                    .as_seconds_f64() as u64
+            }
+            DappToWalletInteractionSubintentExpiration::AfterDelay(delay) => {
+                delay.expire_after_seconds
+            }
+        }
+    }
+
+    /// Given the current epoch and seconds until expiration, returns the end epoch and the max proposer timestamp.
+    fn calculate_end_ranges(
+        &self,
+        current_epoch: Epoch,
+        expiry_time_from_now_in_seconds: u64,
+    ) -> (Epoch, Instant) {
+        // 5 minutes
+        const EXPECTED_EPOCH_DURATION_IN_SECONDS: u64 = 300;
+
+        // As per the transaction validation configuration, epoch diff should be less than 1 month.
+        const MAX_EPOCH_DIFF: u64 =
+            30 * 24 * 60 * 60 / EXPECTED_EPOCH_DURATION_IN_SECONDS;
+
+        // 1 epoch for the fact that it's min_inclusive and max_exclusive; 1 more for the fact that we might be very close to the end of the epoch already
+        const MIN_EPOCH_DIFF: u64 = 2;
+
+        let epoch_diff = MAX_EPOCH_DIFF.min(
+            MIN_EPOCH_DIFF
+                + (expiry_time_from_now_in_seconds
+                    / EXPECTED_EPOCH_DURATION_IN_SECONDS),
```

Comment:
This is good, maybe we should use MIN_EPOCH_DIFF only if:
```rust
(expiry_time_from_now_in_seconds / EXPECTED_EPOCH_DURATION_IN_SECONDS) < MIN_EPOCH_DIFF
```
?

Reply:
I thought about this as well, but I concluded we want to always add 2 more epochs than the epochs until expiration. 

Let's imagine we are in `Epoch(1)` and we have an expiration of 14 minutes. If we apply the suggested change, the result would be `Epoch(3)`. Imagine we are 2 minutes away from next epoch, that means that the user will have the following time period to sign the subintent:
```
Epoch 1: 2 minutes
Epoch 2: 5 minutes
Epoch 3: 5 minutes
```

This is, a total of 12 minutes, while the expiration is in 14 minutes. That's why I believe @dhedey always adds the 2 extra epochs, which he justifies with the constant comment:
```Rust 
// 1 epoch for the fact that it's min_inclusive and max_exclusive; 
// 1 more for the fact that we might be very close to the end of the epoch already
```